app/controllers/objects_controller.rb

```diff
@@ -15,8 +15,18 @@ def show
   # queue a ValidateMoab job for a specific druid, typically called by a preservationIngestWF robot
   # GET /v1/objects/:id/validate_moab
   def validate_moab
-    ValidateMoabJob.perform_later(druid)
-    render(plain: 'ok', status: :ok)
+    # ActiveJob::Base.perform_later will return an instance of the job if it was added to the
+    # queue successfully. It will return false if the job was not enqueued.
+    # The most likely cause is a temporary queue lock for the job and its specific payload, so
+    # it is up to the caller to determine whether retrying later is appropriate.
+    if ValidateMoabJob.perform_later(druid)
+      render(plain: 'ok', status: :ok)
+    else
+      err_msg = "Failed to enqueue ValidateMoabJob for #{druid}. "\
+                'The most likely cause is that the job was already enqueued and is waiting to be picked up. '\
+                'Retry later if appropriate.'
+      render(plain: err_msg, status: :locked)
+    end
```

Comment:
This looks awesome ... wonder if this affects ... other robots?   Other queuing systems besides resque?  Etc.  If it's just about resque ... maybe it should say? 

And how sure are you about this ?

Reply:
> wonder if this affects ... other robots?

I don't think so... since this method (AFAIK) is only invoked by the new `validate-moab` step's robot.

>  Other queuing systems besides resque?

This uses ActiveJob behavior, and so shouldn't be specific to resque.  I mean, practically, I think we'd have trouble switching from e.g. resque to sidekiq because moab-versioning caches things in class vars (iirc), which isn't thread safe (and so isn't sidekiq safe).  But this, at least, shouldn't block switching 🙂

> And how sure are you about this ?

Very sure that this behavior will work in rails 6.1.x (based on lots of testing in rails console in the course of other testing for this PR).  And after some research yesterday, also fairly certain this will work in rails 6.2+...

The `new_framework_defaults_6_0.rb` file has a config entry `Rails.application.config.active_job.return_false_on_aborted_enqueue = true`.  This actually generates a deprecation warning, `DEPRECATION WARNING: return_false_on_aborted_enqueue= is deprecated and will be removed from Rails 6.2`.  this was worrisome, but after some research, what i found indicated that the old default was `false`, and that the old behavior was to return an instance of the job regardless of whether it was successfully queued.  since we had not previously inspected or acted on the output of our `ActiveJob::Base.perform_later` calls, we were able to switch what will be the default behavior in rails 6.2 onward, when we upgraded to rails 6 in dec 2019.

see:
* https://guides.rubyonrails.org/v6.0.3/configuring.html#configuring-active-job (end of section)
* https://github.com/rails/rails/blob/v6.0.4/activejob/CHANGELOG.md#rails-600beta1-january-18-2019
* https://guides.rubyonrails.org/v6.2.0/configuring.html#configuring-active-job (no longer mentions the config field, also note this redirects to the `edgeguides.` URL since 6.2.x is the current latest)

i believe the deprecation warning is because you'll no longer be able to configure the old behavior in 6.2+.

might be good to get confirmation from e.g. @jcoyne on this, since he did the upgrade to rails 6 on this app (iirc), and since he's generally more up on this sort of rails behavioral detail than i am.

Reply:
example from random testing, where i attempt to re-queue a checksum validation job for the same moab before the first job for that moab finished working:

```
[18] pry(main)> ChecksumValidationJob.perform_later(cm)
acquired lock on lock:ChecksumValidationJob-gid://preservation-catalog/CompleteMoab/25315 (none existed)
Enqueued ChecksumValidationJob (Job ID: 37bb8216-40a2-4d6d-97ce-45dac35dabb0) to Resque(checksum_validation) with arguments: #<GlobalID:0x0000000005a45ce0 @uri=#<URI::GID gid://preservation-catalog/CompleteMoab/25315>>
=> #<ChecksumValidationJob:0x0000000005ac8550
 @arguments=
  [#<CompleteMoab:0x00000000062ba178
    id: 25315,
    version: 2,
    preserved_object_id: 25301,
    moab_storage_root_id: 18,
    created_at: Wed, 11 Mar 2020 16:04:37.670398000 UTC +00:00,
    updated_at: Thu, 30 Apr 2020 16:56:11.688153000 UTC +00:00,
    last_moab_validation: Wed, 11 Mar 2020 16:04:37.664436000 UTC +00:00,
    last_checksum_validation: Thu, 30 Apr 2020 16:56:11.686353000 UTC +00:00,
    size: 609026338,
    status: "ok",
    last_version_audit: Thu, 30 Apr 2020 16:56:11.686309000 UTC +00:00,
    status_details: nil,
    from_moab_storage_root_id: nil>],
 @exception_executions={},
 @executions=0,
 @job_id="37bb8216-40a2-4d6d-97ce-45dac35dabb0",
 @priority=nil,
 @queue_name="checksum_validation",
 @timezone="UTC">
[19] pry(main)> ChecksumValidationJob.perform_later(cm)
failed to acquire lock on lock:ChecksumValidationJob-gid://preservation-catalog/CompleteMoab/25315, because it has not expired (1625188536 <= 1625192135)
Failed enqueuing ChecksumValidationJob to Resque(checksum_validation), a before_enqueue callback halted the enqueuing execution.
=> false
```

Reply:
I was wondering about common-accessioning, which definitely uses a lot of perform-later and also uses resque.   Less so about within prescat.